lib/ansible/constants.py

```diff
@@ -103,20 +104,49 @@ def _get_config(p, section, key, env_var, default):
     return default
 
 def load_config_file():
-    ''' Load Config File order(first found is used): ENV, CWD, HOME, /etc/ansible '''
+    ''' Load Config File order(first found is used): ENV, CWD, playbook sibling traversed, HOME, /etc/ansible '''
 
     p = configparser.ConfigParser()
 
+    path_list = []
+
+    # Get config file location from ENV
     path0 = os.getenv("ANSIBLE_CONFIG", None)
     if path0 is not None:
         path0 = os.path.expanduser(path0)
         if os.path.isdir(path0):
             path0 += "/ansible.cfg"
+    path_list.append(path0)
+
+    # Get config file location from CWD
+    # FIXME: Needs deprecation? See: https://github.com/ansible/ansible/issues/11175#issuecomment-109386699
     path1 = os.getcwd() + "/ansible.cfg"
+    path_list.append(path1)
+
+    # It only makes sense to search for playbook location when we use ansible-playbook and we have argv[1]
+    if len(sys.argv[0]) >= 2 and sys.argv[0].endswith("ansible-playbook"):
+        # Only check for playbook reference in position 1
+        # See https://github.com/ansible/ansible/issues/11175#issuecomment-240129682
+        first_argument = sys.argv[1]
+        if os.path.isfile(first_argument):
+            adjacent_path = os.path.dirname(os.path.abspath(first_argument))
+            while True:
+                if adjacent_path == "/":
+                    path_list.append("/ansible.cfg")
+                    break
+                else:
+                    path_list.append(adjacent_path + "/ansible.cfg")
+                adjacent_path = os.path.dirname(adjacent_path)
+
+    # Get config for HOME
     path2 = os.path.expanduser("~/.ansible.cfg")
+    path_list.append(path2)
+
+    # Get config from /etc/ansible
     path3 = "/etc/ansible/ansible.cfg"
+    path_list.append(path3)
 
-    for path in [path0, path1, path2, path3]:
+    for path in path_list:
         if path is not None and os.path.exists(path):
             try:
                 p.read(path)
```

Comment:
Reading the linked feature report, I think this the spec we're trying to implement: https://github.com/ansible/ansible/issues/11175#issuecomment-110485820

So we're missing a warning here if the directory is world writable.   I think that looks something like:

``` python
import stat
[...]
cfg_directory = os.path.dirname(path)
cfg_directory_stat = os.stat(cfg_directory)
if cfg_directory_stat[stat.ST_MODE] & stat.S_IWOTH:
    display.warning('Reading ansible config file from a world readable directory (%s).  This can be a security risk.  If this is the intended location, please change permissions on the directory.' % cfg_directory)